### PR TITLE
📱[5/n] Null Accounts: Update Device Code Flow

### PR DESCRIPTION
**Pull request by kyle-rader-msft**
# PreReqs

Review #261 

## Update Device Code
Device code was following the cached auth pattern fairly well.

### CODE REVIEW

File: src/MSALWrapper.Test/AuthFlow/DeviceCodeTest.cs [hunk 1]
```diff
@@ -83,68 +84,63 @@ public async Task DeviceCodeAuthFlow_CachedToken()
...
 
         [Test]
-        public async Task DeviceCodeAuthFlow_GetTokenSilent_ReturnsNull()
+        public async Task DeviceCodeAuthFlow_GetTokenSilent_ReturnsNull_ThenSuccess()
         {
-            this.SilentAuthReturnsNull();
-
             this.MockAccount();
+            this.SilentAuthReturnsNull();
+            this.DeviceCodeAuthResult();
```
Comment (kyle-rader-msft) [inline src/MSALWrapper.Test/AuthFlow/DeviceCodeTest.cs:106]: This test is the one with the bug - where if silent auth failed by returning null, we never actually attempted auth.

File: src/MSALWrapper.Test/AuthFlow/DeviceCodeTest.cs [hunk 2]
```diff
@@ -83,68 +84,63 @@ public async Task DeviceCodeAuthFlow_CachedToken()
...
 
         [Test]
-        public async Task DeviceCodeAuthFlow_GetTokenSilent_OperationCanceledException()
+        public async Task DeviceCodeAuthFlow_GetTokenSilent_OperationCanceledException_ThenSuccess()
         {
-            this.SilentAuthTimeout();
-
             this.MockAccount();
+            this.SilentAuthTimeout();
+            this.DeviceCodeAuthResult();
```
Comment (kyle-rader-msft) [inline src/MSALWrapper.Test/AuthFlow/DeviceCodeTest.cs:123]: Again here, if SilentAuth Timed out - we never actually attempted auth.

**Review by reillysiemens — COMMENTED** [review]
Comments from #277 apply. Otherwise looks good.

File: src/MSALWrapper/AuthFlow/DeviceCode.cs [hunk 3]
```diff
@@ -63,56 +64,45 @@ public DeviceCode(ILogger logger, Guid clientId, Guid tenantId, IEnumerable<stri
...
-                {
-                    this.errors.Add(ex);
-                    this.logger.LogDebug($"Silent auth failed, re-auth is required.\n{ex.Message}");
-                    var tokenResult = await TaskExecutor.CompleteWithin(
-                        this.logger,
-                        this.deviceCodeFlowTimeout,
-                        "Get Token using Device Code",
-                        (cancellationToken) => this.pcaWrapper.GetTokenDeviceCodeAsync(
+
+                this.logger.LogWarning($"Device Code Authentication for: {this.promptHint}");
```
Comment (shalinikhare27) [inline src/MSALWrapper/AuthFlow/DeviceCode.cs:84]: Same comment as previous PRs. We should assert these log statements in tests.

**Review by reillysiemens — APPROVED** [review]

**Review by shalinikhare27 — APPROVED** [review]

File: src/MSALWrapper/AuthFlow/DeviceCode.cs (hunk 3 above)
Comment (kyle-rader-msft) [reply src/MSALWrapper/AuthFlow/DeviceCode.cs:83]: Resolving with #282

**Review by Haard30 — APPROVED** [review]

**Pull request merged**